File: backend/API/Application/Features/Bussiness/RequestAdoptions/Validators/CreateRequestAdoption.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using API.Application.Features.Bussiness.RequestAdoptions.Dtos.Private;
using FluentValidation;

namespace API.Application.Features.Bussiness.RequestAdoptions.Validators
{
    public class CreateRequestAdoptionValidator : AbstractValidator<CreateRequestAdoption>
    {
        public CreateRequestAdoptionValidator()
        {
            RuleFor(x => x.PetId)
                .NotEmpty()
                    .WithMessage("La mascota es obligatoria.");

            RuleFor(x => x.UserId)
                .NotEmpty()
                    .WithMessage("El usuario es obligatorio.");

            RuleFor(x => x.HouseType)
                .NotEmpty()
                    .WithMessage("El tipo de vivienda es obligatorio.")
                .MaximumLength(50)
                    .WithMessage("El tipo de vivienda no puede exceder los 50 caracteres.");

            RuleFor(x => x.District)
                .NotEmpty()
                    .WithMessage("El distrito es obligatorio.")
                .MaximumLength(100)
                    .WithMessage("El distrito no puede exceder los 100 caracteres.");

            RuleFor(x => x.Phone)
                .NotEmpty()
                    .WithMessage("El teléfono es obligatorio.")
                .MaximumLength(15)
                    .WithMessage("El teléfono no puede exceder los 15 caracteres.")
                .Matches(@"^[\+]?[0-9\s\-\(\)]{7,15}$")
                    .WithMessage("El teléfono debe tener un formato válido (ej. +51 999 999 999 o 999999999).");

            RuleFor(x => x.Motivation)
                .NotEmpty()
                    .WithMessage("La motivación es obligatoria.")
                .MinimumLength(20)
                    .WithMessage("La motivación debe tener al menos 20 caracteres.")
                .MaximumLength(2000)
                    .WithMessage("La motivación no puede exceder los 2000 caracteres.");

            // Validación personalizada: si tiene otras mascotas, debería describirlas
            // (opcional, si agregas el campo OtherPetsDescription)
            RuleFor(x => x.HasOtherPets)
                .NotNull()
                    .WithMessage("Debe indicar si tiene otras mascotas.");

            RuleFor(x => x.HasChildren)
                .NotNull()
                    .WithMessage("Debe indicar si tiene niños en casa.");

            RuleFor(x => x.AcceptHomeVisit)
                .NotNull()
                    .WithMessage("Debe indicar si acepta una visita domiciliaria.");
        }
    }
}

File: frontend/src/features/business/request-adoptions/dto/dashboard/create-request-adoption.ts
import * as Yup from "yup";

const phoneRegExp = /^(\+?\d{1,3}[- ]?)?\d{7,15}$/;

export const CreateRequestAdoptionSchema = Yup.object().shape({
  petId: Yup.string()
    .uuid("El ID de la mascota debe ser un GUID válido")
    .required("El ID de la mascota es obligatorio"),

  userId: Yup.string()
    .uuid("El ID del usuario debe ser un GUID válido")
    .required("El ID del usuario es obligatorio"),

  houseType: Yup.string()
    .max(50, "El tipo de vivienda no puede exceder los 50 caracteres")
    .required("El tipo de vivienda es obligatorio"),

  hasOtherPets: Yup.boolean().default(false),

  hasChildren: Yup.boolean().default(false),

  acceptHomeVisit: Yup.boolean()
    .oneOf([true], "Debes aceptar la visita domiciliaria para continuar")
    .default(false),

  district: Yup.string()
    .max(100, "El distrito no puede exceder los 100 caracteres")
    .required("El distrito es obligatorio"),

  address: Yup.string().optional(),

  reference: Yup.string().nullable().notRequired(),

  dni: Yup.string()
    .matches(/^\d{8}$/, "El DNI debe tener 8 dígitos")
    .required("El DNI es obligatorio"),

  phone: Yup.string()
    .matches(phoneRegExp, "Ingresa un número de teléfono válido")
    .max(15, "El teléfono no puede exceder los 15 caracteres")
    .required("El número de teléfono es obligatorio"),

  motivation: Yup.string()
    .max(2000, "La motivación no puede exceder los 2000 caracteres")
    .required("La motivación o motivo de adopción es obligatorio"),
});

// Opcional: Tipo inferido directamente desde el esquema de Yup
export type CreateRequestAdoptionDto = Yup.InferType<
  typeof CreateRequestAdoptionSchema
>;
